### PR TITLE
feat(): Add support for enum value deprecation and description

### LINKED ISSUE #1126: Support deprecated enum values
The GraphQL Spec [supports deprecated enum values](http://spec.graphql.org/June2018/#sec--deprecated) and NestJS should too.

## I'm submitting a...

<pre><code>
[ ] Regression 
[ ] Bug report
[x] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead post your question on Stack Overflow.
</code></pre>

## Current behavior
`registerEnumType()` takes an object and enumerates all the values of the object as non-deprecated values of the GraphQL enum.

## Expected behavior
I'm not sure how to implement deprecation support for `registerEnumType()`. Possible options:

1. New function taking an object of objects allowing passing a `deprecationReason` for each enum value.
2. A new option (`deprecationReasons`?) in `EnumOptions` that would take a dictionary of deprecation reasons for a subset of the keys of the enum (like `{ ... deprecationReasons: { [key: string]: string }`)
3. A method to mark an enum value as deprecated (maybe like `markEnumValueDeprecated(enumName: string, key: string, deprecationReason?: string)`?)

I'm very keen to get your feedback on a direction, and I'm happy to help with a PR.

### PR DESCRIPTION
**Pull request by mattleff**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

## What is the current behavior?

Closes #1126 

## What is the new behavior?

Enum values may be marked as deprecated by providing a `deprecationReasons` key in the `EnumOptions` argument of `registerEnumType()`.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information

### CODE REVIEW

**Comment by mattleff** [comment]
@kamilmysliwiec After implementing this I thought of adding support for enum value descriptions. It could be supported in a similar way with a `descriptions` key in `EnumOptions`, but I wondered if `valueDescriptions` would be clearer. If we were to do that I wonder if this PR should be adding `valueDeprecationReasons`, wdyt? I'm good with `deprecationReasons` and `descriptions` if you are.

**Comment by mattleff** [comment]
@kamilmysliwiec Sorry to bother you, but I was wondering if you have any review feedback or if we can merge this.

**Comment by longtomjr** [comment]
I would love to see descriptions added in a similar way if possible.

**Comment by kamilmysliwiec** [comment]
Good point @longtomjr. @mattleff I'm just wondering, what would you think about the following API instead:

```typescript
registerEnumType(Direction, {
  name: 'Direction', // this one is mandatory
  description: 'The basic directions', // this one is optional
  valuesMap: {
    Sideways: {
      description: 'Sideways description',
      deprecationReason: 'Replaced with Left or Right',
    }
  },
});
```

or alternatively:

```typescript
valuesMap: {
  [Direction.Sideways]: {
    description: 'Sideways description',
    deprecationReason: 'Replaced with Left or Right',
  }
},
```

**Comment by longtomjr** [comment]
I think I prefer the first option, since it should work with `valuesMap?: { [K in keyof T]?: string };`, but I don't have a strong preference either way. 
The second option is nicer for automatic refactors, but you should get a compile time error either way if it breaks.

**Comment by mattleff** [comment]
Good point about descriptions. I agree with @longtomjr in preferring the first option. I'll update the PR shortly.

**Comment by mattleff** [comment]
@kamilmysliwiec I've pushed a new commit. I'm not sure what the test failure is. Everything passes for me locally.

**Comment by kamilmysliwiec** [comment]
@mattleff would you like to create a PR to the docs as well? :)

File: lib/schema-builder/metadata/enum.metadata.ts [hunk 1]
```diff
@@ -1,5 +1,16 @@
...
-  ref: object;
+export interface EnumMetadataValuesMapOptions {
+  deprecationReason?: string;
+  description?: string;
+}
+
+export type EnumMetadataValuesMap<T extends object> = Record<
+  keyof T,
+  EnumMetadataValuesMapOptions
+>;
```
Comment (longtomjr) [inline lib/schema-builder/metadata/enum.metadata.ts:9]:
```suggestion
export type EnumMetadataValuesMap<T extends object> = Partial<Record<
  keyof T,
  EnumMetadataValuesMapOptions
>>;
```
This should fix the test. I believe record types by default have the keys set as required.

Comment (kamilmysliwiec) [reply lib/schema-builder/metadata/enum.metadata.ts:9]: Thanks!

**Comment by mattleff** [comment]
@kamilmysliwiec Here's a PR for docs: https://github.com/nestjs/docs.nestjs.com/pull/1543.

**Pull request merged**